File: vermyth/engine/keys.py
from __future__ import annotations

import functools

from vermyth.registry import AspectRegistry
from vermyth.schema import Aspect


@functools.cache
def canonical_aspect_key(aspects: frozenset[Aspect]) -> str:
    order_index = {a: i for i, a in enumerate(AspectRegistry.get().full_order)}
    ordered = sorted(aspects, key=lambda a: order_index.get(a, 10**9))
    return "+".join(a.name for a in ordered)

File: typescript/src/engine/keys.ts
import type { Aspect } from "../schema/aspect.js";
import { AspectRegistry } from "../registry.js";

export function canonicalAspectKey(aspects: ReadonlySet<Aspect>): string {
  const order = AspectRegistry.get().fullOrder;
  const orderIndex = new Map(order.map((a, i) => [a.name, i]));
  const ordered = [...aspects].sort(
    (a, b) => (orderIndex.get(a.name) ?? 1e9) - (orderIndex.get(b.name) ?? 1e9),
  );
  return ordered.map((a) => a.name).join("+");
}
